src/photon/photon_scheduler.c

```diff
@@ -53,6 +56,184 @@ void print_resource_info(const local_scheduler_state *state,
 #endif
 }
 
+/**
+ * Kill a worker, if it is a child process, and clean up all of its associated
+ * state.
+ *
+ * @param worker A pointer to the worker we want to kill.
+ * @param wait A bool representing whether we should wait for the worker's
+ *        process to exit. If the worker is not a child process, this flag is
+ *        ignored.
+ * @return Void.
+ */
+void kill_worker(local_scheduler_client *worker, bool wait) {
+  /* Erase the worker from the array of workers. */
+  local_scheduler_state *state = worker->local_scheduler_state;
+  int num_workers = utarray_len(state->workers);
+  for (int i = 0; i < utarray_len(state->workers); ++i) {
+    local_scheduler_client *active_worker =
+        *(local_scheduler_client **) utarray_eltptr(state->workers, i);
+    if (active_worker == worker) {
+      utarray_erase(state->workers, i, 1);
+    }
+  }
+  /* Make sure that we erased exactly 1 worker. */
+  CHECKM(!(utarray_len(state->workers) < num_workers - 1),
+         "Found duplicate workers");
+  CHECKM(utarray_len(state->workers) != num_workers,
+         "Tried to kill worker that doesn't exist");
+
+  /* Remove the client socket from the event loop so that we don't process the
+   * SIGPIPE when the worker is killed. */
+  event_loop_remove_file(worker->local_scheduler_state->loop, worker->sock);
+
+  /* If the worker has registered a process ID with us and it's a child
+   * process, use it to send a kill signal. */
+  if (worker->is_child && worker->pid != 0) {
+    kill(worker->pid, SIGKILL);
+    if (wait) {
+      /* Wait for the process to exit. */
+      waitpid(worker->pid, NULL, 0);
+    }
+  }
+
+  /* Clean up the client socket after killing the worker so that the worker
+   * can't receive the SIGPIPE before exiting. */
+  close(worker->sock);
+
+  /* Clean up the task in progress. */
+  if (worker->task_in_progress) {
+    /* TODO(swang): Update the task table to mark the task as lost. */
```

Comment:
Yes, this PR is a little incomplete right now, since workers are only killed during cleanup, when reporting task death doesn't matter anymore. I'm hoping to implement reporting task death properly in a separate PR for task preemption.